main/src/ca/uwaterloo/flix/language/phase/Redundancy.scala

```diff
@@ -390,9 +390,13 @@ object Redundancy {
       val us2 = visitExp(exp2, env0, rc)
 
       // Check for useless pure expressions.
-      if (exp1.eff == Type.Pure)
```

Comment:
Introduce helper predicates:

- isUseless(exp1)
- isUnderAppliedFunction(exp1)

Note isUnderAppliedFunction should check both for purity, but also that the function actually has an effect!